pkg/controller/route/route_controller.go

```diff
@@ -83,16 +83,40 @@ func New(routes cloudprovider.Routes, kubeClient clientset.Interface, nodeInform
 	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "route_controller"})
 
 	rc := &RouteController{
-		routes:           routes,
-		kubeClient:       kubeClient,
-		clusterName:      clusterName,
-		clusterCIDR:      clusterCIDR,
-		nodeLister:       nodeInformer.Lister(),
-		nodeListerSynced: nodeInformer.Informer().HasSynced,
-		broadcaster:      eventBroadcaster,
-		recorder:         recorder,
+		routes:       routes,
+		kubeClient:   kubeClient,
+		clusterName:  clusterName,
+		clusterCIDR:  clusterCIDR,
+		nodeLister:   nodeInformer.Lister(),
+		nodeInformer: nodeInformer.Informer(),
+		broadcaster:  eventBroadcaster,
+		recorder:     recorder,
 	}
 
+	notified := func() {
+		if err := rc.reconcileNodeRoutes(); err != nil {
+			klog.Errorf("Couldn't reconcile node routes: %v when notified", err)
+		}
+	}
+
+	rc.nodeInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch obj.(type) {
+				case *v1.Node:
+					return true
+				default:
+					return false
+				}
+			},
+
+			Handler: cache.ResourceEventHandlerFuncs{
```

Comment:
Handling `OnUpdate` might actually be useful here only if there is a diff in the PodCIDR of a node (pod cidr is immutable IIRC so it would be from `"" -> <some-valid-cidr>`

Reply:
> Handling `OnUpdate` might actually be useful here only if there is a diff in the PodCIDR of a node

It's a good suggestion.

Reply:
I think we should do this over the OnAdd event, it's unlikely the first Add event for a node will have the pod cidr set already.

Reply:
> it's unlikely the first Add event for a node will have the pod cidr set already.

on controller startup, existing nodes will be passed to the OnAdd handler, and are likely to have a podCIDR set

Reply:
Good to know, thanks! 

I think handling OnUpdate events only still makes more sense since we can do a diff against the old object and only reconcile if the pod cidr changed. Otherwise, the node has to wait for the full resync (same as today).

Reply:
> I think handling OnUpdate events only still makes more sense since we can do a diff against the old object and only reconcile if the pod cidr changed

would it make sense to treat an OnAdd event like an update where the old object had an empty podcidr?

Reply:
Maybe I'm missing something here, wouldn't that trigger a route sync on every node on controller restart?

Reply:
> wouldn't that trigger a route sync on every node on controller restart?

I'd expect the controller to check correctness of existing routes/nodes on start. I'm not attached to use of OnAdd to accomplish that. Does the reconcileNodeRoutes() call inside the non-sliding until loop get run immediately on startup or does it wait 5 minutes initially?

Reply:
IIRC, it does run once immediately. If it doesn't we should fix that. I think OnAdd would be ideal if we can change `reconcileNodeRoutes()` to reconcile a single node object instead of having to reconcile all nodes but that'll require a bit more work here 🤔

Reply:
as long as it runs once on startup, that's sufficient